pkg/config/staging_test.go

```diff
@@ -97,3 +92,91 @@ func (s *zeroSuite) TestGhpcStageFunc(c *C) {
 		c.Errorf("diff (-want +got):\n%s", diff)
 	}
 }
+
+func TestPartialEval(t *testing.T) {
+	type test struct {
+		input string
+		want  string
+	}
+
+	// testing partial evaluation of `upper`
+	tests := []test{
+		{`upper("a")`, `"A"`},
+		{`file(upper("a"))`, `file("A")`},
+		{`lower("A") + upper("b")`, `lower("A")+"B"`},
```

Comment:
Shoudn't `lower` function here update it to lowercase in the want string ?

Reply:
> // testing partial evaluation of `upper` 

not `lower`

> ```{`lower("A") + upper("b")`, `lower("A")+"B"`},```

`lower` is not evaluated because it's not needed to evaluate `upper`

> ```{`upper(lower("A"))`, `"A"`},```

`lower` is evaluated because it's needed to evaluate `upper`